### DISCUSSION TITLE
Incompatible types. Found: 'java.io.File', required: 'java.lang.String' Unhandled exception: java.io.FileNotFoundException Unhandled exception: java.io.IOException

### DISCUSSION THREAD

**Discussion by ghost** [community/community#125711] [Programming Help]
### Body

```java
import java.io.*;
import java.sql.*;
public class SQLManager{
	public boolean isConnectedToDB, passwordRequired;
	public String password, Password, exception;
	public void getPassword(String fileLocation){
		this.Password = new File(fileLocation);
		FileReader fileReader = new FileReader(this.Password);
		this.password = fileReader.read();
		fileReader.close();
		fileReader = null;
	}
	public String enterPassword(String PassWord) throws Exception{
		try{
			if(PassWord == null || this.password == null){
				throw new Exception("`Either input = null` and/or `password = null` is not allowed!\n");
			}else{
				if(PassWord == this.password){
					this.Password = PassWord;
				}else{
					throw new Exception (PassWord.toUpperCase() + " is not the correct password!\n");
				}
			}
		}catch(Error error){
			this.exception = error.toString();
		}
		return "Error: " + this.exception;
	}
	public String connectToDB() throws Exception{
		try{
			if(!this.isConnectedToDB && !this.passwordRequired){
				this.isConnectedToDB = true;
			}else if(!this.isConnectedToDB && this.passwordRequired){
				if(this.Password == this.password) {
					this.isConnectedToDB = true;
				}else{
					throw new Exception (this.Password.toUpperCase() + " is incorrect!\n");
				}
			}else{
				throw new Exception ("You've already connected to the database!\n");
			}
		}catch(Error err) {
			this.exception = err.toString();
		}
		return "Error thrown: " + this.exception.toString();
	}
	public String disconnectFromDB() throws Exception{
		try{
			if(this.isConnectedToDB) {
				this.isConnectedToDB = false;
			}else{
				throw new Exception ("You've already disconnected from the database! \n");
			}
		}catch(Error err){
			this.exception = err.toString();
		}
		return "Error thrown: " + this.exception.toString();
	}
}
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by judahpaul16** [comment]
Just rewrite in Rust. 🤣 But no seriously there are several issues your code. Here's a breakdown of the problems and how to fix them:

1. You are trying to assign a `File` object to a `String` variable. This occurs in the line `this.Password = new File(fileLocation);`. You should maintain the type consistency by keeping `Password` as a `File` and using another `String` variable to read the contents of the file.

2. You are catching generic `Error` instead of `Exception`. It's a good practice to catch more specific exceptions like `IOException` for file operations.

3. You are comparing strings using `==`, which checks for reference equality rather than content equality. Use `.equals()` for string comparison.

4. Java typically uses camel case for variable names (`password` instead of `Password`). Also, avoid using multiple variables with similar names (`password` and `Password`) as it can lead to confusion.

Here's a revised version of your `SQLManager` class that addresses these issues:

```java
import java.io.*;
import java.sql.*;

public class SQLManager {
    public boolean isConnectedToDB, passwordRequired;
    public String password; // Changed to maintain type consistency
    public File passwordFile; // File object for the password file
    public String exception;

    public void getPassword(String fileLocation) throws IOException {
        passwordFile = new File(fileLocation);
        FileReader fileReader = new FileReader(passwordFile);
        BufferedReader bufferedReader = new BufferedReader(fileReader);
        password = bufferedReader.readLine(); // Reads the first line from the file as password
        bufferedReader.close();
        fileReader.close();
    }

    public String enterPassword(String inputPassword) throws Exception {
        if (inputPassword == null || password == null) {
            throw new Exception("`inputPassword = null` and/or `password = null` is not allowed!\n");
        } else {
            if (inputPassword.equals(password)) {
                passwordFile = new File(inputPassword); // Questionable usage
            } else {
                throw new Exception(inputPassword.toUpperCase() + " is not the correct password!\n");
            }
        }
        return "Password accepted";
    }

    public String connectToDB() throws Exception {
        if (!isConnectedToDB && !passwordRequired) {
            isConnectedToDB = true;
        } else if (!isConnectedToDB && passwordRequired) {
            if (new File(password).equals(passwordFile)) {
                isConnectedToDB = true;
            } else {
                throw new Exception(new String(password).toUpperCase() + " is incorrect!\n");
            }
        } else {
            throw new Exception("You've already connected to the database!\n");
        }
        return "Connection established";
    }

    public String disconnectFromDB() throws Exception {
        if (isConnectedToDB) {
            isConnectedToDB = false;
        } else {
            throw new Exception("You've already disconnected from the database!\n");
        }
        return "Disconnected successfully";
    }
}
```

There's still some ambiguity and questionable logic in the way `passwordFile` is used, especially in the `enterPassword` method where setting it as a new `File` might not be meaningful. Could you provide more details about the use case?

**Comment by PALLADIUM26** [comment]
Hi, can you please specify at which line the specified error is occurring? 

Again in line 18, we cannot check equality of strings using '==' operator. Instead, you can use .equals() function.
Syntax: `if(PassWord.equals(this.password))`

In the catch block we should catch Exception not error.
Syntax: `catch(Exception err){ //statements }`

I hope you will find this useful.

**Comment by ignjatic01** [comment]
In the line this.Password = new File(fileLocation);, you're trying to assign a File object to a String type (Password). This is causing the Incompatible types error. Instead, you should read the content of the file as a String.

You are using FileReader without handling the FileNotFoundException and IOException. These exceptions should be either handled or declared in the method signature.

//Note
Always use equals(Object o) method for objects (== is only for primitive types)

**Comment by EbbeWertzStudentUH** [comment]
It's this section:

`this.Password = new File(fileLocation);
FileReader fileReader = new FileReader(this.Password);`

new File is just a constructor call on file, so this will output a File Object. Now you try to put it into the `password` member of your class. But the password member is a String object and not a File Object.
What you can do to fix this, is simply initialise a new file variable:
`File fileThatContainsPassword = new File(location);`
Then you pass this `fileThatContainsPassword` into the filereader instead of `this.password`.
This fileThatContainsPassword can also just be a local variable in the getPassword method since you don't need the file object itself anywhere else. Also, this fileread may throw an exception, so it's best to surround this with a try catch. Specifically for a filereader, a better approach can also be a _try with resources_, which will automatically close the file when it's finished using it.

Additionally, your other methods throw exceptions which they also catch themselves. This is redundant.
If you want your class to throw exceptions, you just need to keep the throws Exception specifier next to your method names. All try catch blocks inside are not needed.
If you want to handle your errors internally in this class, you don't need the throw exception on the method signature. This is because you already have a try catch block that catches the exception that yout throw inside the function.
Also if you intentionally throw your exception, it has no use to catch it again, since catching is meanth to prevent an actual program-stopping error and just do some action like returning a string. Since this catch catches an intentional exception your throw yourself, it has no use, and you can achieve the same result by directly returning the error message string from the function without working with any execeptions.

A third approach is, when you want errors to occur, but don't want to be forced to use try catch blocks when calling your methods is to throw runtime exceptions. Again, this does not need a try catch block inside the method since this defeats the purpose of throwing an exception, but you also don't need the throws Exception on the method signature since a runtime exception can be used to throw sudden, "unexpected" exceptions. If you use this approach be carefull though since your program will "crash" when the exception is not caught, especially given the fact that a runtime exception also doesn't force cathcing.
